src/NServiceBus.Core/Performance/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -0,0 +1,56 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Metrics;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class ReceiveDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
+    {
+        static readonly Counter<long> TotalProcessedSuccessfully =
+            MessagingMetricsFeature.NServiceBusMeter.CreateCounter<long>("messaging.successes", description: "Total number of messages processed successfully by the endpoint.");
+
+        static readonly Counter<long> TotalFetched =
+            MessagingMetricsFeature.NServiceBusMeter.CreateCounter<long>("messaging.fetches", description: "Total number of messages fetched from the queue by the endpoint.");
+
+        static readonly Counter<long> TotalFailures =
+            MessagingMetricsFeature.NServiceBusMeter.CreateCounter<long>("messaging.failures", description: "Total number of messages processed unsuccessfully by the endpoint.");
+
+        public ReceiveDiagnosticsBehavior(string queueNameBase, string discriminator)
+        {
+            this.queueNameBase = queueNameBase;
+            this.discriminator = discriminator;
+        }
+
+        public async Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
+        {
+            context.MessageHeaders.TryGetValue(Headers.EnclosedMessageTypes, out var messageTypes);
+
+            var tags = new List<KeyValuePair<string, object>>
+            {
+                new KeyValuePair<string, object>("messaging.discriminator", discriminator ?? ""),
```

Comment:
are these tag names coming from some spec or are those arbitrary names? should we put them in the `nservicebus `namespace instead of `messaging`?

Reply:
I think it makes sense. Happy to raise a small PR after this is merged.